### PR TITLE
[CORE] Skip unnecessary convert_to_supported_precision if ConstantFolding is omitted

### PR DESCRIPTION
**Pull request by itikhono**
Details:
It's a modification of https://github.com/openvinotoolkit/openvino/pull/22674
f16 LLM (llama was tested) compilation time on ARM is unreasonable huge. Perf report shows that every ConstantFolding transformation takes several seconds even if the graph is not modified.
The root cause is util::convert_to_supported_precision call even if constant folding is skipped.
The suggested fix is to skip util::convert_to_supported_precision call if folding is not applied.

Tickets:
CVS-152428

### CODE REVIEW

File: src/core/src/op/assign.cpp [hunk 1]
```diff
@@ -137,6 +137,10 @@ bool Assign::has_evaluate() const {
 bool Assign::constant_fold(OutputVector& output_values, const OutputVector& inputs_values) {
```
Comment (praasz) [inline src/core/src/op/assign.cpp:137]: Can be removed as Base implementation will return false.

File: src/core/src/op/constant.cpp [hunk 2]
```diff
@@ -658,6 +658,10 @@ bool Constant::constant_fold(OutputVector&, const OutputVector&) {
     return false;
```
Comment (praasz) [inline src/core/src/op/constant.cpp:658]: Can be removed as Base implementation will return false.

File: src/core/src/op/random_uniform.cpp [hunk 3]
```diff
@@ -92,6 +92,10 @@ bool RandomUniform::constant_fold(OutputVector& output_values, const OutputVecto
     return false;
```
Comment (praasz) [inline src/core/src/op/random_uniform.cpp:92]: Can be removed as Base implementation will return false.

File: src/core/src/op/read_value.cpp [hunk 4]
```diff
@@ -179,6 +179,10 @@ bool ReadValue::has_evaluate() const {
 bool ReadValue::constant_fold(OutputVector& output_values, const OutputVector& inputs_values) {
```
Comment (praasz) [inline src/core/src/op/read_value.cpp:179]: Can be removed as Base implementation will return false.

File: src/core/src/op/result.cpp [hunk 5]
```diff
@@ -71,6 +71,10 @@ bool Result::constant_fold(OutputVector& output_values, const OutputVector& inpu
     return false;
```
Comment (praasz) [inline src/core/src/op/result.cpp:71]: Can be removed as Base implementation will return false.

File: src/core/include/openvino/core/node.hpp [hunk 6]
```diff
@@ -207,6 +207,7 @@ class OPENVINO_API Node : public std::enable_shared_from_this<Node> {
     virtual bool evaluate_upper(ov::TensorVector& output_values) const;
     virtual bool evaluate_symbol(TensorSymbolVector& output_symbols) const;
 
+    virtual bool can_constant_fold(const OutputVector& inputs_values);
```
Comment (praasz) [inline src/core/include/openvino/core/node.hpp:210]:
```suggestion
    virtual bool can_constant_fold(const OutputVector& inputs_values) const;
```

**Comment by itikhono** [comment]
this PR is under perf validation checks

**Comment by itikhono** [comment]
another possible fix of this issue: https://github.com/openvinotoolkit/openvino/pull/27086

File: src/core/src/op/squeeze.cpp [hunk 7]
```diff
@@ -104,9 +104,16 @@ bool Squeeze::evaluate_symbol(TensorSymbolVector& output_symbols) const {
     return validate::axes_has_and_set_bound(*this) && ov::util::default_symbol_evaluator(this, output_symbols);
 }
 
+bool Squeeze::can_constant_fold(const OutputVector& inputs_values) {
+    if (get_output_partial_shape(0).is_dynamic() || is_const_fold_disabled()) {
+        return false;
+    }
+    return true;
```
Comment (CuriousPanCake) [inline src/core/src/op/squeeze.cpp:111]:
```suggestion
    return get_output_partial_shape(0).is_static() && !is_const_fold_disabled();
```

File: src/core/src/op/util/gather_base.cpp [hunk 8]
```diff
@@ -262,13 +255,22 @@ bool GatherBase::evaluate_symbol(TensorSymbolVector& output_symbols) const {
     return gather::have_indices_and_axis_bound_set(this) && ov::util::default_symbol_evaluator(this, output_symbols);
 }
 
+bool GatherBase::can_constant_fold(const OutputVector& input_values) {
+    if (get_output_partial_shape(0).is_dynamic() || input_values.size() != 3) {
+        return false;
+    }
+    return true;
```
Comment (CuriousPanCake) [inline src/core/src/op/util/gather_base.cpp:262]:
```suggestion
    return get_output_partial_shape(0).is_static() && input_values.size() == 3;
```

Comment (itikhono) [reply src/core/src/op/util/gather_base.cpp]: done

File: src/core/src/op/squeeze.cpp (hunk 7 above)
Comment (itikhono) [reply src/core/src/op/squeeze.cpp]: done

File: src/core/src/op/result.cpp (hunk 5 above)
Comment (itikhono) [reply src/core/src/op/result.cpp]: done

File: src/core/src/op/read_value.cpp (hunk 4 above)
Comment (itikhono) [reply src/core/src/op/read_value.cpp]: done

File: src/core/src/op/random_uniform.cpp (hunk 3 above)
Comment (itikhono) [reply src/core/src/op/random_uniform.cpp]: done

File: src/core/src/op/assign.cpp (hunk 1 above)
Comment (itikhono) [reply src/core/src/op/assign.cpp]: done

**Comment by CuriousPanCake** [comment]
LGTM. I will approve after performance validation

**Review by CuriousPanCake — APPROVED** [review]

**Comment by itikhono** [comment]
> LGTM. I will approve after performance validation

We already have the perf validation results on CPU.
The compile time was improved up to 18-20% for specific models.
The results are attached to the ticket.

**Review by praasz — APPROVED** [review]

**Pull request merged**